### ISSUE TITLE
Check for shape consistency between model and guide

### ISSUE THREAD

**Issue by lumip** [pyro-ppl/numpyro#1119]
In SVI latent random variables sampled from the guide completely mask those sampled in the model during inference. However, nothing prevents us from specifying different shapes for such a sample site in model and guide respectively. This makes it easy to introduce confusing bugs when code in the model expects a certain shape that is different than what guide provides. This can easily happen if the model needs an adaptation but the user forgets to reflects that in the guide.

The problem then is not as much that this results in an error, but where that error occurs: In the code following the sampling of the RV in the model (when the actual error might have been sampling a wrong shape in the guide). Furthermore, the model function itself is sane, i.e., if the user searches for the error there, they will not find anything, thus leading to much confusion.

Long story short, I think it would be beneficial to check that shapes of sample sites referring to the same random variable in guide and model are consistent and at least issue a warning (if not outright raise an error) if this is not the case, as that would make it much easier to spot this type of error. Thoughts?

**Comment by fritzo** [comment]
@fehiepsi have you you already ported Pyro's [check_model_guide_match()](https://github.com/pyro-ppl/pyro/blob/7e293c5627458948db100c6a12d1e566f9f45327/pyro/util.py#L238) to NumPyro?

**Comment by fehiepsi** [comment]
Currently, we don't have that check in numpyro. This should be a good first issue. :)

**Comment by maw501** [comment]
Hey @fehiepsi ,

I could possibly look into this if you think it's suitable for someone new to the library (and contributing to open source in general 😬 ).

Alternatively if you think there are any other issues better suited for me to look into (e.g. #951) then I'm happy to take your suggestion. 😄 

Mark

**Comment by fehiepsi** [comment]
Hi @maw501, sure, please go ahead. If you are new to the library, it is best to start with [tensor shapes in Pyro](https://pyro.ai/examples/tensor_shapes.html). You might also want to play with [trace handler](http://num.pyro.ai/en/stable/handlers.html#trace), which you can use the inspect your model. Please let me know if you have any question during the way.

(#951 is also a good first issue - both issues do not require a deep understanding of the library)
